internal/service/connect/instance_storage_config.go

```diff
@@ -0,0 +1,525 @@
+	"github.com/hashicorp/terraform-provider-aws/internal/verify"
+)
+
+func ResourceInstanceStorageConfig() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceInstanceStorageConfigCreate,
+		ReadContext:   resourceInstanceStorageConfigRead,
+		UpdateContext: resourceInstanceStorageConfigUpdate,
+		DeleteContext: resourceInstanceStorageConfigDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"association_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"instance_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringLenBetween(1, 100),
+			},
+			"resource_type": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice(connect.InstanceStorageResourceType_Values(), false),
+			},
+			"storage_config": {
+				Type:     schema.TypeList,
+				Required: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"kinesis_firehose_config": {
+							Type:     schema.TypeList,
+							Optional: true,
+							MaxItems: 1,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"firehose_arn": {
+										Type:         schema.TypeString,
+										Required:     true,
+										ValidateFunc: verify.ValidARN,
+									},
+								},
+							},
+						},
+						"kinesis_stream_config": {
+							Type:     schema.TypeList,
+							Optional: true,
+							MaxItems: 1,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"stream_arn": {
+										Type:         schema.TypeString,
+										Required:     true,
+										ValidateFunc: verify.ValidARN,
+									},
+								},
+							},
+						},
+						"kinesis_video_stream_config": {
+							Type:     schema.TypeList,
+							Optional: true,
+							MaxItems: 1,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"encryption_config": {
+										Type:     schema.TypeList,
+										Required: true,
+										MaxItems: 1,
+										Elem: &schema.Resource{
+											Schema: map[string]*schema.Schema{
+												"encryption_type": {
+													Type:         schema.TypeString,
+													Required:     true,
+													ValidateFunc: validation.StringInSlice(connect.EncryptionType_Values(), false),
+												},
+												"key_id": {
+													Type:         schema.TypeString,
+													Required:     true,
+													ValidateFunc: verify.ValidARN,
+												},
+											},
+										},
+									},
+									// API returns <prefix>-connect-<connect_instance_alias>-contact-
+									"prefix": {
+										Type:         schema.TypeString,
+										Required:     true,
+										ValidateFunc: validation.StringLenBetween(1, 128),
+									},
```

Comment:
For the `prefix` field, the handling of the difference between what the practitioner provides and what the AWS API returns should be handled by either a `StateFunc` or a `SuppressDiffFunc`. If the returned value of `prefix` could be used, e.g., to reference other resources, the full value should be stored and a `SuppressDiffFunc` used to ignore differences. Otherwise, a `StateFunc` could be used to modify the value returned by the AWS API before any operations.

Reply:
Resolved in 5da39fd7eb09a288c5f3a3509575fe94ad66a36e